### PR TITLE
WIP: convert stacked coords to OrderedDict.

### PR DESCRIPTION
**Pull request by mlshapiro**
This is probably 75% done, but has resulted in some unintended consequences:

- `StackedCoordinates` now requires dims/names on creation, otherwise we can't keep track. I could use some index values as names (i.e. 0, 1), but that would be strange if they were never set. I can't use None (like the other implementing BaseCoordinates) since we have multiple dimensions using the same None key overwriting each other. For now, I required name. 
- Because of the name requirements, `clinspace` is not as intuitive for stacked coordinates. I removed `clinspace` support for `StackedCoordinates`. 
- `StackedCoordinates` end up looking like a mixture between `Coordinates` and `BaseCoordinates`. I think this is intuitive, but it may not be since its really multi-indexed.

Note I removed `read_only`, but after talking with @mpu-creare I will add back in.

At the end of the day, I want the stacked coordinates to be intuitive. Before I go deeper lets decide if this is the right path.

Alternatively, I can just change the `StackedCoordinates._coords` type to be a list, which will let me implement the transform work. That will have less consequences, though I've already gone through the effort to implement this, I feel like its worth considering.

### CODE REVIEW

File: podpac/core/coordinates/cfunctions.py [hunk 1]
```diff
@@ -36,18 +36,18 @@ def crange(start, stop, step, name=None):
 
 def clinspace(start, stop, size, name=None):
```
Comment (jmilloy) [inline podpac/core/coordinates/cfunctions.py]: I think the stacked linspace is important, and I think we can use the name argument here to support it. Something like this

```
    ...
    dims = name.split('_')
    if np.array(start).size != len(dims):
        raise size mismatch

    a = np.array([start, stop])
    if a.ndim == 2:
        c = StackedCoordinates([UniformCoordinates1d(start[i], stop[i], size=size) for i, dim in enumerate(dims)])
    else:
        c = UniformCoordinates1d(start, stop, size=size, name=name)
    return c
```

File: podpac/core/coordinates/coordinates.py [hunk 2]
```diff
@@ -784,7 +802,7 @@ def udrop(self, dims, ignore_missing=False):
                 if c.name not in dims:
                     cs.append(c)
             elif isinstance(c, StackedCoordinates):
-                stacked = [s for s in c if s.name not in dims]
+                stacked = [s for s in c.values() if s.name not in dims]
```
Comment (jmilloy) [inline podpac/core/coordinates/coordinates.py]: Let's try to use either `c.values()` or `c._coords.values()` throughout; I think we have both at the moment.

File: podpac/core/coordinates/coordinates.py [hunk 3]
```diff
@@ -110,8 +114,20 @@ def __init__(self, coords, dims=None, coord_ref_sys=None, ctype=None, distance_u
             if isinstance(coords[i], BaseCoordinates):
                 c = coords[i]
             elif '_' in dim:
-                cs = [val if isinstance(val, Coordinates1d) else ArrayCoordinates1d(val) for val in coords[i]]
-                c = StackedCoordinates(cs)
+                # make sure components of stacked coordinates have dim names from stacked dim
```
Comment (jmilloy) [inline podpac/core/coordinates/coordinates.py]: Hmmm. This is messy.

Is this a new check that we weren't doing before (to check for a name mismatch with existing Coordinates1d), or is this just because we need to set it now (instead of later) so we have to the check now?

I think you also need to check that `len(sdims) == len(coords[i])`.

File: podpac/core/coordinates/coordinates.py [hunk 4]
```diff
@@ -866,7 +884,14 @@ def unique(self):
             New Coordinates object with unique, sorted coordinate values in each dimension.
         """
 
-        return Coordinates([c[np.unique(c.coordinates, return_index=True)[1]] for c in self.values()])
+        coords = list(self.values())
```
Comment (jmilloy) [inline podpac/core/coordinates/coordinates.py]: What happened here? I'm curious what didn't work about the previous version.

File: podpac/core/coordinates/stacked_coordinates.py [hunk 5]
```diff
@@ -222,15 +227,42 @@ def copy(self):
             Copy of the stacked coordinates, with provided properties and name.
         """
 
-        return StackedCoordinates(self._coords)
+        return StackedCoordinates(list(self._coords.values()))
```
Comment (jmilloy) [inline podpac/core/coordinates/stacked_coordinates.py]: This is a sort of a good argument *against* using a dictionary for _coords.

File: podpac/core/coordinates/stacked_coordinates.py [hunk 6]
```diff
@@ -240,14 +272,11 @@ def __iter__(self):
     def __len__(self):
         return len(self._coords)
 
-    def __getitem__(self, index):
-        if isinstance(index, string_types):
-            if index not in self.dims:
-                raise KeyError("Dimension '%s' not found in dims %s" % (index, self.dims))
-            return self._coords[self.dims.index(index)]
+    def __getitem__(self, dim):
```
Comment (jmilloy) [inline podpac/core/coordinates/stacked_coordinates.py]: Ahh. I use the indexing I had here a lot. Why did you remove this part

```
    else:
        return StackedCoordinates([c[index] for c in self._coords])
```

File: podpac/core/coordinates/coordinates.py (hunk 4 above)
Comment (jmilloy) [reply podpac/core/coordinates/coordinates.py]: Ahh, I see, the indexing is removed. I think you need to put that back, and then you can revert this.

**Comment by jmilloy** [comment]
I'm sort of lukewarm on this. It seems like it simplifies `__getitem__` and `__setitem__` *slightly* and complicates initialization *slightly*. If you could accomplish the same thing by just changing the tuple to a list, I think that would make a lot of sense.

File: podpac/core/coordinates/stacked_coordinates.py [hunk 7]
```diff
@@ -260,16 +289,39 @@ def __setitem__(self, dim, c):
...
+            return False
+
+        return True
+
+    def __delitem__(self, dim):
+        if not dim in self.dims:
+            raise KeyError("Cannot delete dimension '%s' in StackedCoordinates %s" % (dim, self.dims))
 
-        self.set_trait('_coords', coords)
+        del self._coords[dim]
```
Comment (jmilloy) [inline podpac/core/coordinates/stacked_coordinates.py]: I think you need to check that this doesn't leave the StackedCoordinates with only one dimension.

File: podpac/core/coordinates/stacked_coordinates.py [hunk 8]
```diff
@@ -222,15 +227,42 @@ def copy(self):
...
         """
 
-        return StackedCoordinates(self._coords)
+        return StackedCoordinates(list(self._coords.values()))
 
     # ------------------------------------------------------------------------------------------------------------------
-    # standard methods, tuple-like
+    # standard dict-like methods
     # ------------------------------------------------------------------------------------------------------------------
+    def keys(self):
```
Comment (jmilloy) [inline podpac/core/coordinates/stacked_coordinates.py]: I probably wouldn't include `keys`, `items`, `get` and `update` in the public StackedCoordinates interface. It does make sense to have `values` in the public interface, though I would probably leave `iter` with the original meaning (now implemented as `return self._coords.values()`) and then you can use that instead `values`. But I think that's a small thing. I like

```
c = StackedCoordinates([<lat>, <lon>])
lat, lon = c
```

But I can see that this might be clearer.

```
c = StackedCoordinates([<lat>, <lon>])
lat, lon = c.values()
```

**Comment by mlshapiro** [comment]
@jmilloy i greatly appreciate the view - i went with changing the `_coords` to being a list and that works, and it also makes intuitive sense.  i'm going to leave this through the release then we'll likely prune this branch
